src/common/middleware/call-api.js

```diff
@@ -0,0 +1,101 @@
+import { checkStatus, getError } from '../helpers/api';
+
+// Action key that carries API call info interpreted by this Redux middleware.
+export const CALL_API = 'CALL_API';
+
+// A Redux middleware that interprets actions with CALL_API info specified.
+// Performs the call and promises when such actions are dispatched.
+export default defaults => () => next => async action => {
+  // If action does not invoke CALL_API,
+  // ignore it
+  const settings = action[CALL_API];
+  if (typeof settings === 'undefined') {
+    return next(action);
+  }
+
+  // Validate action settings
+  validateDefaults(defaults);
+  validateSettings(settings);
+
+  const [ successType, failureType ] = settings.types;
+  const resource = defaults.endpoint + settings.path;
+  const optionsWithCsrfToken = withCsrfToken(defaults.csrfToken, settings.options);
+  const createAction = createActionWith.bind({}, action);
+
+  // If action has a type, dispatch the action next
+  if (action.type) {
+    next(createAction());
+  }
+
+  try {
+    const response = await fetch(resource, optionsWithCsrfToken);
+    await checkStatus(response);
+    const result = await response.json();
+
+    if (result.status !== 'success') {
+      throw getError(response, result);
+    }
+
+    return next(createAction({
+      type: successType,
+      response: result.payload
+    }));
+  }
+  catch(error) {
+    return next(createAction({
+      type: failureType,
+      error: error
```

Comment:
@katiefenn that could work, but actual payload will get quite ugly nested.

Would it work as something like this?
```
{
  type: 'ACTION_TYPE',
  payload: {
    ...actionPayload: { /* payload from original action */ }
    ...callApiPayload: { /* response or error from callApi just added to original payload */ },
  }
}
```

Of course there would be a risk of callApi payload overriding the action payload - but it overrides action type anyway when calling SUCCESS/ERROR.

The other way maybe:
```
{
  type: 'ACTION_TYPE',
  payload: {
    ...actionPayload: { /* payload from original action */ }
    response: callApiPayload { /* response or error from callApi nested */ },
  }
}
```

So that only response object would be nested? And this would be API response object (whatever is returned by API) in case of success or error object in case of error?

These are just quick proposals, I'm not sure exactly how it would match specific bits of callApi middleware.